File: tmobile-poptoken-lib/lib-tmobile-oss-poptoken-builder/src/main/java/com/tmobile/oss/security/taap/poptoken/builder/PopTokenBuilder.java
package com.tmobile.oss.security.taap.poptoken.builder;

import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.UUID;

import org.apache.commons.codec.binary.Base64;
import org.apache.commons.codec.digest.DigestUtils;
import org.apache.commons.lang3.NotImplementedException;
import org.apache.commons.lang3.StringUtils;

import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.interfaces.RSAKeyProvider;
import com.tmobile.oss.security.taap.poptoken.builder.exception.PopTokenBuilderException;
import com.tmobile.oss.security.taap.poptoken.builder.utils.PopTokenBuilderUtils;

/**
 * A builder class to build the PoP token.
 * 
 * PopTokenBuilder is not thread-safe so the instance of this class should not be shared between multiple threads.
 */
public class PopTokenBuilder {

    protected static final int POP_TOKEN_VALIDITY_DURATION_IN_MILLIS = 2 * 60 * 1000;
    protected static final int MAX_NUMBER_OF_EHTS = 100;

    private RSAPrivateKey rsaPrivateKey;
    private String rsaPrivateKeyPemString;
    private String privateKeyPassword;
    private LinkedHashMap<String, String> ehtsKeyValueMap;

    /**
     * Returns the new instance of PopTokenBuilder.
     * 
     * @return The new instance of PopTokenBuilder
     */
    public static PopTokenBuilder newInstance() {
        return new PopTokenBuilder();
    }

    /**
     * Sets the ehtsKeyValueMap, this LinkedHashMap will be used to calculate the values of ehts (external headers to sign) and edts
     * (hash of external data to sign).
     * 
     * @param ehtsKeyValueMap The map containing the values for all the ehts (external headers to sign) keys.
     *            <p>
     *            Determining the ehts key name:
     *            <ul>
     *            <li>For HTTP request URI, "uri" should be used as ehts key name, <code>PopEhtsKey.URI.keyName()</code>. <br>
     *            For "uri" ehts value, URI and query string of the request URL should be put in the map. Example: If the URL is
     *            "https://api.t-mobile.com/commerce/v1/orders?account-number=0000000000" then only
     *            "/commerce/v1/orders?account-number=0000000000" should be used as ehts "uri" value. The query parameter values part
     *            of "uri" ehts should not be in URL encoded format.</li>
     *            <li>For HTTP method, "http-method" should be used as ehts key
     *            name,<code>PopEhtsKey.HTTP_METHOD.keyName()</code>.</li>
     *            <li>For HTTP request headers, the header name should be used as ehts key name.</li>
     *            <li>For HTTP request body, "body" should be used as ehts key name, <code>PopEhtsKey.BODY.keyName()</code>.</li>
     *            </ul>
     * @return The PopTokenBuilder
     */
    public PopTokenBuilder setEhtsKeyValueMap(LinkedHashMap<String, String> ehtsKeyValueMap) {
        this.ehtsKeyValueMap = ehtsKeyValueMap;
        return this;
    }

    /**
     * Signs the PoP token using the specified rsaPrivateKeyPemString.
     * 
     * @param rsaPrivateKeyPemString The RSA private key PEM string
     * @return The PopTokenBuilder
     */
    public PopTokenBuilder signWith(String rsaPrivateKeyPemString) {
        this.rsaPrivateKeyPemString = rsaPrivateKeyPemString;
        return this;
    }

    /**
     * Signs the PoP token using the specified encryptedRsaPrivateKeyPemString and uses the privateKeyPassword to decrypt the private
     * key PEM string string.
     * 
     * @param encryptedRsaPrivateKeyPemString The encrypted RSA private key PEM string
     * @param privateKeyPassword The RSA private key password
     * @return The PopTokenBuilder
     */
    public PopTokenBuilder signWith(String encryptedRsaPrivateKeyPemString, String privateKeyPassword) {
        this.rsaPrivateKeyPemString = encryptedRsaPrivateKeyPemString;
        this.privateKeyPassword = privateKeyPassword;
        return this;
    }

    /**
     * Signs the PoP token using the specified RSAPrivateKey.
     * 
     * @param rsaPrivateKey The RSA private key
     * @return The PopTokenBuilder
     */
    public PopTokenBuilder signWith(RSAPrivateKey rsaPrivateKey) {
        this.rsaPrivateKey = rsaPrivateKey;
        return this;
    }

    /**
     * Returns the issued at time.
     * <p>
     * Note: This method is having the default access for JUnit tests to override it by creating a sub class of PopTokenBuilder.
     * 
     * @return The issued at time
     */
    Date getIssuedAt() {
        return new Date();
    }

    /**
     * Returns the expiration time.
     * 
     * @param issuedAt The time when the PoP token is issued
     * @return The expiration time
     */
    protected Date getExpiration(Date issuedAt) {
        return new Date(issuedAt.getTime() + POP_TOKEN_VALIDITY_DURATION_IN_MILLIS);
    }

    /**
     * Returns the version.
     * 
     * @return The version
     */
    protected String getVersion() {
        return "1";
    }

    /**
     * Returns the unique identifier.
     * 
     * @return The unique identifier
     */
    protected String getUniqueIdentifier() {
        return UUID.randomUUID().toString();
    }

    /**
     * Builds, signs and returns the string representation of the PoP token. The PoP token will be valid for 2 minutes.
     * 
     * @return The PoP token
     * @throws PopTokenBuilderException if error occurs while building the PoP token
     */
    public String build() throws PopTokenBuilderException {

        try {
            if (ehtsKeyValueMap == null || ehtsKeyValueMap.isEmpty() || doesContainAnyEmptyKeysOrValues(ehtsKeyValueMap)) {
                throw new PopTokenBuilderException(
                        "The ehtsKeyValueMap should not be null or empty and should not contain any null or empty ehts keys or values");
            }

            if (ehtsKeyValueMap.size() > 100) {
                throw new PopTokenBuilderException(
                        "The ehtsKeyValueMap should not contain more than " + MAX_NUMBER_OF_EHTS + " entries");
            }

            if (rsaPrivateKey == null && StringUtils.isBlank(rsaPrivateKeyPemString)) {
                throw new PopTokenBuilderException(
                        "Either rsaPrivateKey or rsaPrivateKeyPemString should be provided to sign the PoP token");
            }

            if (rsaPrivateKey != null && StringUtils.isNotBlank(rsaPrivateKeyPemString)) {
                throw new PopTokenBuilderException(
                        "Either only rsaPrivateKey or only rsaPrivateKeyPemString should be provided to sign the PoP token");
            }

            if (rsaPrivateKey == null) {
                if (privateKeyPassword != null) {
                    rsaPrivateKey = PopTokenBuilderUtils.encryptedKeyPemStringToRsaPrivateKey(rsaPrivateKeyPemString,
                            privateKeyPassword);
                } else {
                    rsaPrivateKey = PopTokenBuilderUtils.keyPemStringToRsaPrivateKey(rsaPrivateKeyPemString);
                }
            }

            Date issuedAt = getIssuedAt();

            return JWT.create() //
                    .withClaim("ehts", buildEhtsString(ehtsKeyValueMap)) //
                    .withClaim("edts", calculateEdtsSha256Base64Hash(ehtsKeyValueMap)) //
                    .withClaim("jti", getUniqueIdentifier()) //
                    .withClaim("v", getVersion()) //
                    .withIssuedAt(issuedAt) //
                    .withExpiresAt(getExpiration(issuedAt)) //
                    .sign(buildRsa256Algorithm(rsaPrivateKey, null)); //

        } catch (PopTokenBuilderException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new PopTokenBuilderException("Error occurred while building the PoP token, error: " + ex.toString(), ex);
        }
    }

    // ===== helper methods ===== //

    /**
     * Returns true if the ehtsKeyValueMap contains any empty keys
     * 
     * @param ehtsKeyValueMap
     * @return
     */
    private boolean doesContainAnyEmptyKeysOrValues(LinkedHashMap<String, String> ehtsKeyValueMap) {
        for (Entry<String, String> ehtsKeyValueEntry : ehtsKeyValueMap.entrySet()) {
            if (StringUtils.isBlank(ehtsKeyValueEntry.getKey()) || StringUtils.isBlank(ehtsKeyValueEntry.getValue())) {
                return true;
            }
        }
        return false;
    }

    /**
     * Builds the semicolon separated ehts string using the specified ehtsKeysList
     * 
     * @param ehtsKeyValueMap The ehts keys list
     * @return the ehts string containing the semicolon delimited ehts keys
     */
    private String buildEhtsString(LinkedHashMap<String, String> ehtsKeyValueMap) {

        StringBuilder ehtsStringBuilder = new StringBuilder();
        for (String ehtsKey : ehtsKeyValueMap.keySet()) {
            if (ehtsStringBuilder.length() > 0) {
                ehtsStringBuilder.append(";");
            }
            ehtsStringBuilder.append(ehtsKey);
        }
        return ehtsStringBuilder.toString();
    }

    /**
     * Calculates the edts (hash of external data to sign) in the same sequence as provided in the LinkedHashMap.
     * 
     * @param ehtsKeysList
     * @param ehtsKeyValueMap
     * @return The edts (hash of external data to sign)
     */
    private String calculateEdtsSha256Base64Hash(Map<String, String> ehtsKeyValueMap) {
        StringBuilder stringToHashBuilder = new StringBuilder();
        for (Entry<String, String> ehtsKeyValueEntry : ehtsKeyValueMap.entrySet()) {
            stringToHashBuilder.append(ehtsKeyValueEntry.getValue());
        }
        String stringToHash = stringToHashBuilder.toString();
        return sha256Base64UrlSafeString(stringToHash);
    }

    /**
     * Creates SHA 256 base64 URL safe encrypted hash of the specified string.
     * 
     * @param stringToHash The string to be hashed
     * @return SHA 256 base64 encrypted hash.
     */
    private String sha256Base64UrlSafeString(String stringToHash) {
        byte[] sha256Digest = DigestUtils.sha256(stringToHash);
        return Base64.encodeBase64URLSafeString(sha256Digest);
    }

    /**
     * Builds the RSA 256 algorithm.
     * 
     * @param rsaPrivateKey The RSA private key
     * @param privateKeyId The private key ID
     * @return The RSA 256 algorithm
     */
    private Algorithm buildRsa256Algorithm(RSAPrivateKey rsaPrivateKey, String privateKeyId) {
        Algorithm algorithm = Algorithm.RSA256(new RSAKeyProvider() {
            @Override
            public RSAPublicKey getPublicKeyById(String keyId) {
                throw new NotImplementedException("The method getPublicKeyById is not implemented");
            }

            @Override
            public RSAPrivateKey getPrivateKey() {
                return rsaPrivateKey;
            }

            @Override
            public String getPrivateKeyId() {
                return privateKeyId;
            }
        });
        return algorithm;
    }
}

File: poptoken-lib/poptoken-builder/cs-lib-tmobile-oss-poptoken-builder/com.tmobile.oss.security.taap.poptoken.builder/PopTokenBuilder.cs
/*
 * Copyright 2020 T-Mobile US, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using Microsoft.IdentityModel.Tokens;
using Microsoft.IdentityModel.JsonWebTokens;
using System.Runtime.CompilerServices;
using System;
using System.Security.Cryptography;
using System.Text;
using System.Collections.Generic;
using System.Security.Claims;

[assembly: InternalsVisibleTo("com.tmobile.oss.security.taap.poptoken.builder.test")]
namespace com.tmobile.oss.security.taap.poptoken.builder
{
    /// <summary>
    /// A builder class to build the PoP token. PopTokenBuilder is not thread-safe. 
    /// So the instance of this class should not be shared between multiple threads.
    /// </summary>
    public class PopTokenBuilder : IPopTokenBuilder
    {
        private const int POP_TOKEN_VALIDITY_DURATION_IN_MILLIS = 2 * 60 * 1000;
        private const int MAX_NUMBER_OF_EHTS = 100;
        private const string POP_TOKEN_VERSION = "1";
        private string audience;
        private string issuer;
        private HashSet<KeyValuePair<string, string>> ehtsKeyValueMap;

        internal string privateKeyXmlRsa;
        internal RsaSecurityKey rsaSecurityKey;

        /// <summary>
        /// Default Constructor
        /// </summary>
        private PopTokenBuilder()
        {
        }

        /// <summary>
        /// Custom Constructor
        /// </summary>
        public PopTokenBuilder(string audience, string issuer) : this()
        {
            this.audience = audience;
            this.issuer = issuer;
        }

        /// <summary>
        /// Returns the Issued At DateTime in UTC.
        /// </summary>
        /// <returns>The issued at time</returns>
        public DateTime IssuedAt
        {
            get
            {
                return DateTime.UtcNow;
            }
        }

        /// <summary>
        /// Returns the version.
        /// </summary>
        /// <returns>The version</returns>
        public string Version
        {
            get
            {
                return POP_TOKEN_VERSION;
            }
        }

        /// <summary>
        /// Sets the ehtsKeyValueMap, which is a HashSet of KeyValuePair. 
        /// It is used to calculate the values of ehts (external headers to sign) and edts (hash of external data to sign).
        /// </summary>
        /// <param name="ehtsKeyValueMap">HashSet of headers and data</param>
        /// <returns>The PopTokenBuilder</returns>
        public PopTokenBuilder SetEhtsKeyValueMap(HashSet<KeyValuePair<string, string>> ehtsKeyValueMap)
        {
            this.ehtsKeyValueMap = ehtsKeyValueMap;
            return this;
        }

        /// <summary>
        /// Signs the PoP token using the specified rsaSecurityKeyPemString.
        /// </summary>
        /// <param name="privateKeyXmlRsa">The RSA private key in XML format</param>
        /// <returns>The PopTokenBuilder</returns>
        public PopTokenBuilder SignWith(string privateKeyXmlRsa)
        {
            this.privateKeyXmlRsa = privateKeyXmlRsa;
            return this;
        }

        /// <summary>
        /// Returns the expiration time.
        /// </summary>
        /// <param name="issuedAt"> The time when the PoP token is issued </param>
        /// <returns> The expiration time </returns>
        public DateTime GetExpiration(DateTime issuedAt)
        {
            return issuedAt.AddMilliseconds((double)POP_TOKEN_VALIDITY_DURATION_IN_MILLIS);
        }

        /// <summary>
        /// Returns the unique identifier.
        /// </summary>
        /// <returns> The unique identifier </returns>
        public string GetUniqueIdentifier()
        {
            return Guid.NewGuid().ToString("D");
        }

        /// <summary>
        /// Builds, signs and returns the string representation of the PoP token. The PoP token will be valid for 2 minutes.
        /// </summary>
        /// <returns> The PoP token </returns>
        /// <exception cref="PopTokenBuilderException"> if error occurs while building the PoP token </exception>
        public string Build()
        {
            try
            {
                if (this.ehtsKeyValueMap == null ||
                    this.ehtsKeyValueMap.Count == 0 ||
                    DoesContainAnyEmptyKeysOrValues(this.ehtsKeyValueMap))
                {
                    throw new PopTokenBuilderException("The ehtsKeyValueMap should not be null or empty and should not contain any null or empty ehts keys or values");
                }

                if (this.ehtsKeyValueMap.Count > MAX_NUMBER_OF_EHTS)
                {
                    throw new PopTokenBuilderException("The ehtsKeyValueMap should not contain more than " + MAX_NUMBER_OF_EHTS + " entries");
                }

                if (this.rsaSecurityKey == null &&
                    string.IsNullOrEmpty(this.privateKeyXmlRsa))
                {
                    throw new PopTokenBuilderException("Either rsaSecurityKey or privateKeyXmlRsa should be provided to sign the PoP token");
                }

                var rsa = RSA.Create();
                rsa.KeySize = 2048;
                rsa.FromXmlRsaPemKey(this.privateKeyXmlRsa);
                this.rsaSecurityKey = new RsaSecurityKey(rsa);

                var ehts = BuildEhtsString(this.ehtsKeyValueMap);
                var edts = CalculateEdtsSha256Base64Hash(this.ehtsKeyValueMap);
                var jti = GetUniqueIdentifier();
                var version = this.Version;
                var issuedAt = this.IssuedAt;
                var expires = GetExpiration(issuedAt);
                var signingCredentials = new SigningCredentials(this.rsaSecurityKey, SecurityAlgorithms.RsaSha256);
                var securityTokenDescriptor = new SecurityTokenDescriptor()
                {
                    Subject = new ClaimsIdentity(new List<Claim>
                    {
                        new Claim("ehts", ehts),
                        new Claim("edts", edts),
                        new Claim("v", version),
                        new Claim("jti", jti)
                    }),
                    Audience = this.audience,
                    Issuer = this.issuer,
                    NotBefore = issuedAt,
                    IssuedAt = issuedAt,
                    Expires = expires,
                    SigningCredentials = signingCredentials
                };

                var jsonWebTokenHandler = new JsonWebTokenHandler();
				return jsonWebTokenHandler.CreateToken(securityTokenDescriptor); // Any instance members are not guaranteed to be thread safe.
            }
            catch (PopTokenBuilderException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new PopTokenBuilderException("Error occurred while building the PoP token builder, error: " + ex.Message, ex);
            }
        }

        /// <summary>
        /// Returns true if the ehtsKeyValueMap contains any empty keys
        /// </summary>
        /// <param name="ehtsKeyValueMap">
        /// @return </param>
        private bool DoesContainAnyEmptyKeysOrValues(HashSet<KeyValuePair<string, string>> ehtsKeyValueMap)
        {
            foreach (var ehtsKeyValueEntry in ehtsKeyValueMap)
            {
                if (string.IsNullOrEmpty(ehtsKeyValueEntry.Key) ||
                    string.IsNullOrEmpty(ehtsKeyValueEntry.Value))
                {
                    return true;
                }
            }

            return false;
        }

        /// <summary>
        /// Builds the semicolon separated ehts string using the specified ehtsKeysList
        /// </summary>
        /// <param name="ehtsKeyValueMap"> The ehts keys list </param>
        /// <returns> the ehts string containing the semicolon delimited ehts keys </returns>
        private string BuildEhtsString(HashSet<KeyValuePair<string, string>> ehtsKeyValueMap)
        {
            var ehtsStringBuilder = new StringBuilder();

            foreach (var ehtsKey in ehtsKeyValueMap)
            {
                if (ehtsStringBuilder.Length > 0)
                {
                    ehtsStringBuilder.Append(";");
                }

                ehtsStringBuilder.Append(ehtsKey.Key.Replace("\n", "")
                                                    .Replace("\r", ""));
            }

            return ehtsStringBuilder.ToString();
        }

        /// <summary>
        /// Calculates the edts (hash of external data to sign) in the same sequence as provided in the LinkedHashMap.
        /// </summary>
        /// <param name="ehtsKeysList"> </param>
        /// <param name="ehtsKeyValueMap"> </param>
        /// <returns> The edts (hash of external data to sign) </returns>
        private string CalculateEdtsSha256Base64Hash(HashSet<KeyValuePair<string, string>> ehtsKeyValueMap)
        {
            var stringToHashBuilder = new StringBuilder();

            foreach (var ehtsKeyValueEntry in ehtsKeyValueMap)
            {
                stringToHashBuilder.Append(ehtsKeyValueEntry.Value
                                                            .Replace("\n", "")
                                                            .Replace("\r", ""));
            }

            var stringToHash = stringToHashBuilder.ToString();

            return Sha256Base64UrlSafeString(stringToHash);
        }

        /// <summary>
        /// Creates SHA 256 base64 URL safe encrypted hash of the specified string.
        /// </summary>
        /// <param name="stringToHash"> The string to be hashed </param>
        /// <returns> SHA 256 base64 encrypted hash. </returns>
        private string Sha256Base64UrlSafeString(string stringToHash)
        {
            var inputBytes = Encoding.ASCII.GetBytes(stringToHash);
            var sha256 = new SHA256Managed();
            var hash = sha256.ComputeHash(inputBytes);

            return Convert.ToBase64String(hash).Replace("=", "")
                                               .Replace("+", "-")
                                               .Replace("/", "_");
        }
    }
}
